### DISCUSSION TITLE
Use ImbalancedSampler on node level for full dataset?

### DISCUSSION THREAD

**Discussion by udiallo** [pyg-team/pytorch_geometric#5836] [Q&A]
Hi,

I have multiple graphs with unbalanced classes on which I attempt to train a model on a node multi-class classification task. I would like to apply the ImbalancedSampler (https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/loader/imbalanced_sampler.html) to the Dataset object, consisting of multiple graphs (as Data objects). As far as I understand the documentation, currently I can either do sampling on node level with a single Data object, or graph-level sampling by passing a Dataset object. In the second case, the ImbalancedSampler throws an error in case that the number of classes is not equal to the number of graphs in the Dataset (which makes sense since it is meant for graph classification I guess).

Is there a way to handle my inbalanced graphs while training for a node classification task? Can I adapt the ImbalancedSampler in a way to that the sampler is applied than for all nodes in all graphs? Or are there other workarounds for unbalanced classes in multiple graphs while node classification?

Thank you in advance.

**Comment by wsad1** [comment]
You can convert n graphs into a single graph with n disconnected components and then pass that to `ImbalancedSampler`.
```
# this converts your dataset into a single data object
# data.batch will tell you which graph each node belongs to.
data = torch_geometric.Batch.from_data_list(dataset[:])
sampler = ImbalancedSampler(data, input_nodes=data.train_mask)
```

**Comment by rusty1s** [comment]
I don't think you can really do imbalanced sampling on node-level tasks across multiple graphs. How should that work? As an alternative, I think it may be good to simply re-weight your loss. PyTorch provides native support for this via the `weight` argument in all its loss functions.

**Reply by udiallo** [reply]
Okay, so I have up to a hundreds graphs (might get more) labeled with 11 different classes (each node of each graph is labeled with one of 11 labels, however in some graphs it might be the case that some classes are not found at all). Currently the model tends to only predict 2 or 3 of these labels and ignores the others completely. This makes sense as these labels are very overrepresented in the graph. So that's why I thought it might makes sense (also following the GraphSAGE idea) to use some kind of sampling in order to use the same amount of nodes from each class while training.

Can you elaborate again on why this won't work? Using the massive single but disconnected graph mentioned in the above answer, it can be seen as a masking of overrepresented nodes while training, but without the need of removing nodes from the graph (which is actually not possible, since this would influence the other nodes as well).

I will try to use the weight argument of the loss function. Would you set the weights for the single classes here to the opposite of the class distribution? So in case I have 5 classes with` [1000, 1000, 1000000, 1000, 1000] `samples (extreme example, I know), the weight vector would look something like `[1.0, 1.0, 0.002, 1.0, 1.0]` ?

Thank you for your quick responses.

**Reply by rusty1s** [reply]
Yes, sure. Moving to a supergraph and using `NeighborLoader` works as well, but this will be ultimately slower.

[This](https://discuss.pytorch.org/t/passing-the-weights-to-crossentropyloss-correctly/14731) thread here goes into more detail on setting `weight`. IMO, it is best to set it to something like `[1.0, 1.0, 0.002, 1.0, 1.0]`.
